### PR TITLE
HHH-19497 Test for composition of 'not' 'in' restrictions

### PR DESCRIPTION
**Pull request by michaelfranz**
not in is not handled correctly for multi-column joins. This pull request fixes the problem.

Compositions of not with in on lists fail for dialects not supporting record-level construction, such as DB2Dialect.
Using not with in with such dialects can lead to dangerous outcomes, such as all records being returned from a query. This is because the query condition always evaluates to either true or false (depending on the details of the query) because multi-column joins are not handled correctly.

----------------------
By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt)
and can be relicensed under the terms of the [LGPL v2.1 license](https://www.gnu.org/licenses/old-licenses/lgpl-2.1.txt) in the future at the maintainers' discretion.
For more information on licensing, please check [here](https://github.com/hibernate/hibernate-orm/blob/main/CONTRIBUTING.md#legal).

----------------------

https://hibernate.atlassian.net/browse/HHH-19497

### CODE REVIEW

**Comment by gavinking** [comment]
@beikov would you take a look at this one please?

File: hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java [hunk 1]
```diff
@@ -7664,14 +7664,17 @@ else if ( !dialect.supportsRowValueConstructorSyntaxInInList() ) {
...
+						appendSql(separator);
 						emulateTupleComparison(
 								lhsTuple.getExpressions(),
-								getSqlTuple( expression ).getExpressions(),
-								comparisonOperator,
+								SqlTupleContainer.getSqlTuple(expression).getExpressions(),
+								ComparisonOperator.EQUAL,
 								true
 						);
 						separator = " or ";
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java]: Thanks for the fix and test. I'm just curious why you didn't change the junction type here instead. Am I missing some subtle case where using `not ( (x=x1 and y=y1) or (x=x2 and y=y2) )` is "more" correct over `( (x<>x1 or y<>y1) and (x<>x2 or y<>y2) )`?

```suggestion
						separator = inListPredicate.isNegated() ? " and " : " or ";
```

I also did some testing with `null` values on PostgreSQL and to me it seems that both kinds of expressions are semantically equivalent. With an increasing amount of in-list items (3+), the `not` approach is more compact, which is why I would even prefer that. Since you seem to be on DB2, could you please let me know if there is a performance difference between the two predicate styles i.e. are the query execution plans the same?

Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java]: To be clear, I don't have a strong preference for one solution or the other. Your previous solution seemed nicer to me, because with a growing number of in-list items, the resulting SQL is simply shorter, but ultimately, I would prefer we go with the solution that is "fastest".

Now, let me ask you again. Since you seem to be on DB2, could you please let me know if there is a performance difference between the two predicate styles i.e. are the query execution plans the same?

Comment (michaelfranz) [reply hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java]: I will happily provide some findings on execution plans in the coming days, project priorities permitting. In the meantime, I again stress that the problem is not db2-specific. It impacts any database (driver) that does not support record-level or tuple construction. Knowing db2's plan is interesting, perhaps, but it's just one case.
Ideally one would survey the most popular databases to see their execution plans and relative performance for this case; and make a decision based on what works best for most.

Comment (michaelfranz) [reply hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java]: As requested I have examined the execution plan under DB2. I did this with the following, logically equivalent queries:

```
SELECT *
FROM sch.BIG_TABLE TT
WHERE 
(TT.T1='2022-06-10-05.31.10.133597' AND TT.T2='2022-06-10-05.31.10.327766')
OR
(TT.T1='2022-06-10-05.31.10.411034' AND TT.T2='2022-06-10-05.31.10.433176')
OR
(TT.T1='2022-06-10-05.31.10.452352' AND TT.T2='2022-06-10-05.31.10.472124')
;

SELECT *
FROM sch.BIG_TABLE tt
WHERE 
NOT (
NOT (TT.T1='2022-06-10-05.31.10.133597' AND TT.T2='2022-06-10-05.31.10.327766')
AND
NOT (TT.T1='2022-06-10-05.31.10.411034' AND TT.T2='2022-06-10-05.31.10.433176')
AND
NOT (TT.T1='2022-06-10-05.31.10.452352' AND TT.T2='2022-06-10-05.31.10.472124')
);


SELECT *
FROM sch.BIG_TABLE tt
WHERE 
NOT (
(TT.T1<>'2022-06-10-05.31.10.133597' OR TT.T2<>'2022-06-10-05.31.10.327766')
AND
(TT.T1<>'2022-06-10-05.31.10.411034' OR TT.T2<>'2022-06-10-05.31.10.433176')
AND
(TT.T1<>'2022-06-10-05.31.10.452352' OR TT.T2<>'2022-06-10-05.31.10.472124')
);


```

The execution plan for each of these came up exactly the same. I mean, literally, zero difference. Here is the execution plan output:


```
| Operator   | Name              | Id | Est. Cardinality | Total Cost |
+------------+-------------------+----+------------------+------------+
| Return     |                   |  1 |            3.000 |     63.466 |
| Fetch      |                   |  2 |    7,841,902.000 |     63.466 |
|            | TA: SCH.BIG_TABLE |    |    7,841,902.000 |            |
| RID Scan   |                   |  3 |            1.000 |     42.319 |
| Sort       |                   |  8 |            1.000 |     14.107 |
| Index scan |                   |  9 |    7,841,902.000 |     14.106 |
|            | IX: SYSIBM.SQL... |    |    7,841,902.000 |            |
| Sort       |                   |  6 |            1.000 |     14.107 |
| Index scan |                   |  7 |    7,841,902.000 |     14.106 |		
|            | IX: SYSIBM.SQL... |    |    7,841,902.000 |            |
| Sort       |                   |  4 |            1.000 |     14.107 |
| Index scan |                   |  5 |    7,841,902.000 |     14.106 |		
|            | IX: SYSIBM.SQL... |    |    7,841,902.000 |            |

```

In conclusion, a modern database will generally derive the same, optimised predicate, for logically equivalent predicates. DB2 is no exception.

Back to the implementation of the solution to this bug: Following @beikov 's hint, I did re-implement the solution in a way that effectively distributes the NOT across predicate clauses. I find this solution better because the resulting Java code is slightly terser. Yet I have no reason to believe this will impact performance on modern database engines.

Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java]: Thanks a lot for posting your findings. I had imagined it would be that way, but it's good to see you confirm my guess.

**Comment by beikov** [comment]
Can you please rebase and squash your commits?

**Comment by michaelfranz** [comment]
> Can you please rebase and squash your commits?

Done.

**Comment by michaelfranz** [comment]
@beikov quick question: Now that this contribution has been merged, what are your thoughts about creating a fix (presumably leading to a minor release) of Hibernate v6.6.x? Again, as a newbie to the Hibernate project, I'm not sure how this is usually done. Thanks.

**Comment by beikov** [comment]
The release is automated and happens on Sunday.
